### PR TITLE
PSScriptAnalyzer

### PR DESCRIPTION
**Pull request by zigford**
I use powershell in Vim a lot and have not written any significant vimscript before, and I saw there was an open issue #1264 for a new linter.

### CODE REVIEW

**Comment by zigford** [comment]
Ps, I’m not sure why this is failing at the moment. Could someone with more knowledge of the tests give me a hint?

File: ale_linters/powershell/psscriptanalyzer.vim [hunk 1]
```diff
@@ -0,0 +1,154 @@
+" Author: Jesse Harris - https://github.com/zigford
+" Description: This file adds support for lintng powershell scripts
+"   using the PSScriptAnalyzer module.
+
+" let g:ale_powershell_psscriptanalyzer_exclusions =
+" \ 'PSAvoidUsingWriteHost,PSAvoidGlobalVars'
+call ale#Set('powershell_psscriptanalyzer_exclusions',
```
Comment (w0rp) [inline ale_linters/powershell/psscriptanalyzer.vim]: You should just write `''` as the second argument here.

File: ale_linters/powershell/psscriptanalyzer.vim [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+\ get(g:,
+\ 'ale_linters_powershell_psscriptanalyzer_exclusions', ''))
+call ale#Set('powershell_psscriptanalyzer_executable', 'pwsh')
+call ale#Set('powershell_psscriptanalyzer_module',
+\ 'psscriptanalyzer')
+
+" GetExecutable {{{
```
Comment (w0rp) [inline ale_linters/powershell/psscriptanalyzer.vim]: Get rid of these `{{{` and `}}}` comments. You can use the indentation levels for folding instead.

File: ale_linters/powershell/psscriptanalyzer.vim [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        return ''
+    endif
+
+    "let l:executable = ale#Escape(l:executable)
+    let l:module_string = ale_linters#powershell#psscriptanalyzer#Escape(
+    \ ' -NoProfile -Command "&{$m=Get-Module -ListAvailable '
+    \ . l:module . ';$m.Version.ToString()}"')
+
+    return l:executable . l:module_string
+endfunction
```
Comment (w0rp) [inline ale_linters/powershell/psscriptanalyzer.vim]: If you aren't using the code, you should remove it.

File: ale_linters/powershell/psscriptanalyzer.vim [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+" return the command used to execute it
+function! s:TemporaryPSScript(buffer, input) abort
+    let l:filename = 'script.ps1'
+
+    " Create a temporary filename, <temp_dir>/<original_basename>
+    " The file itself will not be created by this function.
+    let l:tempscript =
+    \ ale#util#Tempname() . (has('win32') ? '\' : '/') . l:filename
+
+    if ale#command#CreateTempFile(a:buffer, l:tempscript, a:input)
```
Comment (w0rp) [inline ale_linters/powershell/psscriptanalyzer.vim]: Use `ale#command#CreateFile` or `ale#Command#ManageFile` instead. `CreateTempFile` isn't intended for use outside of some very specific parts of code.

Comment (zigford) [reply ale_linters/powershell/psscriptanalyzer.vim]: Thanks for your feedback, implemented your suggestions in commit a839a9e4197101182accdf9cf48867def5479bec

**Comment by w0rp** [comment]
I'll merge this and fix the linter errors.

**Comment by coolaj86** [comment]
@zigford Do you know how / where I specify the `PSScriptAnalyzerSettings.psd1` for my project on Mac / Linux?

I've looked through the source here (though I don't read VimScript very well) and a few PowerShell docs on Microsoft Learn, but I can't figure out how to do it.

This is what I have in the root of my project directory (where I'm running vim from), but the warnings are not being ignored as I expected:

`./PSScriptAnalyzerSettings.psd1`:
```pwsh
@{
    IncludeRules = @()
    ExcludeRules = @(
        'PSAvoidUsingCmdletAliases',
        'PSUseDeclaredVarsMoreThanAssignments'
    )
}
```

When I run `vim ./bat/install.ps1` I still see all the errors.

**Comment by zigford** [comment]
> @zigford Do you know how / where I specify the `PSScriptAnalyzerSettings.psd1` for my project on Mac / Linux?
> 
> I've looked through the source here (though I don't read VimScript very well) and a few PowerShell docs on Microsoft Learn, but I can't figure out how to do it.
> 
> This is what I have in the root of my project directory (where I'm running vim from), but the warnings are not being ignored as I expected:
> 
> `./PSScriptAnalyzerSettings.psd1`:
> 
> ```powershell
> @{
>     IncludeRules = @()
>     ExcludeRules = @(
>         'PSAvoidUsingCmdletAliases',
>         'PSUseDeclaredVarsMoreThanAssignments'
>     )
> }
> ```
> 
> When I run `vim ./bat/install.ps1` I still see all the errors.

I think it’s in the :help but put something like this in your vimrc
  " Suppress Write-Host and Global vars warnings
  let g:ale_powershell_psscriptanalyzer_exclusions = 'PSAvoidUsingWriteHost,PSAvoidGlobalVars'

**Comment by coolaj86** [comment]
I am using `g:ale_powershell_psscriptanalyzer_exclusions` right now, however, my understanding is that I should be able to place a `PSScriptAnalyzerSettings.psd1` in the project root with any and all supported options and have that pass through as `-Settings <project-root>/PSScriptAnalyzerSettings.psd1` just like VS Code and other editors that support PowerShell.

I suppose that wasn't available at the time or otherwise didn't get implemented?

**Pull request merged**